### PR TITLE
feat: Use single db-wide executor for running queries

### PR DESCRIPTION
**Pull request by alamb**
Follow on to https://github.com/influxdata/influxdb_iox/pull/1191, and related to https://github.com/influxdata/conductor/issues/244 where running CPU heavy queries can cause IOx to stop responding to health check queries.

# Rationale:
Ensure that all queries are executed on the dedicated worker pool rather than the main tokio pool

# Changes:
1. Plumb a single `Arc<Executor>` through the Server and down into each Db
2. Ensure that all invocations of planning / plans happen via the `Executor`'s thread pool and not the (implicit) tokio one

# Done:
- [x] Audit code for all explicit instantiations of `Executor` and remove as many as possible
- [x] Audit code for uses of datafusion::physical_plan::collect
- [x] Audit code for use of `tokio::task::spawn`
- [x] Remove some extraneous `async` functions and corresponding `await`

# Not yet done
I plan this for the (next) PR:
- [ ] move planning into the Executor thread pool

### CODE REVIEW

File: arrow_deps/src/test_util.rs [hunk 1]
```diff
@@ -29,6 +29,46 @@ macro_rules! assert_table_eq {
...
+/// This is a macro so errors appear on the correct line
+///
+/// Designed so that failure output can be directly copy/pasted
+/// into the test code as expected results.
+///
+/// Expects to be called about like this:
+///
+/// `assert_batch_sorted_eq!(expected_lines: &[&str], batches: &[RecordBatch])`
+#[macro_export]
+macro_rules! assert_batches_sorted_eq {
```
Comment (alamb) [inline arrow_deps/src/test_util.rs:43]: I cribbed this from DataFusion (but, I also actually contributed it to DataFusion in the first place.... )

File: server/src/db.rs [hunk 2]
```diff
@@ -845,7 +855,7 @@ mod tests {
             "+-----+------+",
         ];
         let batches = run_query(Arc::clone(&db), "select * from cpu").await;
-        assert_table_eq!(expected, &batches);
+        assert_batches_sorted_eq!(expected, &batches);
```
Comment (alamb) [inline server/src/db.rs:858]: now that the batches are run on a separate thread pool the output order can be different in the test (aka the output is no longer deterministic)

File: server/src/lib.rs [hunk 3]
```diff
@@ -273,20 +273,15 @@ impl<M: ConnectionManager> Server<M> {
     }
 
     /// Tells the server the set of rules for a database.
-    pub async fn create_database(
-        &self,
-        rules: DatabaseRules,
-        server_id: NonZeroU32,
-        object_store: Arc<ObjectStore>,
-    ) -> Result<()> {
+    pub async fn create_database(&self, rules: DatabaseRules, server_id: NonZeroU32) -> Result<()> {
```
Comment (alamb) [inline server/src/lib.rs:276]: the `Server` already has a `object_store` so there is no need to pass it in again here

File: query/src/exec/context.rs [hunk 4]
```diff
@@ -169,14 +188,21 @@ impl IOxExecutionContext {
...
-        if physical_plan.output_partitioning().partition_count() <= 1 {
-            physical_plan.execute(0).await
-        } else {
-            // merge into a single partition
-            let plan = MergeExec::new(physical_plan);
-            // MergeExec must produce a single partition
-            assert_eq!(1, plan.output_partitioning().partition_count());
-            plan.execute(0).await
-        }
+        self.exec
```
Comment (alamb) [inline query/src/exec/context.rs:191]: I will be honest and say I am not entire sure about this change (aka how a stream of futures will work on a separate executor). For now, the only thing that uses this happens to be running on the separate thread pool anyways

File: query/src/exec.rs [hunk 5]
```diff
@@ -271,11 +278,10 @@ impl Executor {
             .into_iter()
             .map(|plan| {
                 let ctx = self.new_context();
-                // TODO run these on some executor other than the main tokio pool
```
Comment (alamb) [inline query/src/exec.rs:274]: hey look, I got rid of a todo I had put in there a while ago!

File: query/src/exec/context.rs [hunk 6]
```diff
@@ -146,13 +146,13 @@ impl IOxExecutionContext {
 
     /// Prepare a SQL statement for execution. This assumes that any
     /// tables referenced in the SQL have been registered with this context
-    pub async fn prepare_sql(&mut self, sql: &str) -> Result<Arc<dyn ExecutionPlan>> {
+    pub fn prepare_sql(&mut self, sql: &str) -> Result<Arc<dyn ExecutionPlan>> {
```
Comment (alamb) [inline query/src/exec/context.rs:149]: these were some leftover `async` that I removed while auditing the code

File: server/src/db.rs [hunk 7]
```diff
@@ -223,8 +223,12 @@ pub struct Db {
 
     pub server_id: NonZeroU32, // this is also the Query Server ID
 
+    /// Interface to use for peristence
     pub store: Arc<ObjectStore>,
 
+    /// Executor for running queries
+    exec: Arc<Executor>,
```
Comment (alamb) [inline server/src/db.rs:230]: this is the core of this PR: threading this `Executor` all the way through to query execution

File: arrow_deps/src/test_util.rs [hunk 8]
```diff
@@ -29,6 +29,46 @@ macro_rules! assert_table_eq {
     };
 }
 
+/// Compares formatted output of a record batch with an expected
+/// vector of strings in a way that order does not matter.
```
Comment (NGA-TRAN) [inline arrow_deps/src/test_util.rs:33]: Useful tests. I see myself use this a lot in the future

File: query/src/exec.rs (hunk 5 above)
Comment (NGA-TRAN) [reply query/src/exec.rs:274]: Yay!

File: server/src/lib.rs (hunk 3 above)
Comment (NGA-TRAN) [reply server/src/lib.rs:285]: +1

File: server/src/db.rs (hunk 7 above)
Comment (NGA-TRAN) [reply server/src/db.rs:230]: Yay, the DB now use the thread pool it is provided!

**Review by NGA-TRAN — APPROVED** [review]
Looks great, Andrew. Now we can manage thread pool used for queries better!

**Pull request merged**